patches/flashtools.patch

```diff
@@ -0,0 +1,302 @@
++		fprintf(stderr, "Failed to read header\n");
++		return EXIT_FAILURE;
++	}
++	header.magic = ntohl(header.magic);
++	header.version = ntohl(header.version);
++	header.romsize = ntohl(header.romsize);
++	header.bootblocksize = ntohl(header.bootblocksize);
++	header.align = ntohl(header.align);
++	header.offset = ntohl(header.offset);
++	header.architecture = ntohl(header.architecture);
++
++	if (verbose) {
++		printf("Header magic          : %x\n", header.magic);
++		printf("Header version        : %x\n", header.version);
++		printf("Header ROM size       : %x\n", header.romsize);
++		printf("Header boot block size: %x\n", header.bootblocksize);
++		printf("Header align          : %x\n", header.align);
++		printf("Header offset         : %x\n", header.offset);
++		printf("Header arch           : %x\n", header.architecture);
++	}
++
++	if (header.magic != CBFS_HEADER_MAGIC) {
++		fprintf(stderr, "Failed to find valid header\n");
++		return EXIT_FAILURE;
++	}
++
++	uint32_t align = header.align;
++	// loop through files
++	uint64_t off = end - ((uint64_t) header.romsize) + ((uint64_t) header.offset);
++	while (1) {
++		if (verbose) {
++			printf("Potential CBFS File Offset: %lx\n", off);
++		}
++	  if (lseek(fd, off, SEEK_SET) == -1) {
++	    fprintf(stderr, "Failed to seek to next CBFS offset %lx\n", off);
++	    return EXIT_FAILURE;
++	  }
++
++		struct cbfs_file file;
++		if (read(fd, &file, sizeof(struct cbfs_file)) < 0) {
++			fprintf(stderr, "Failed to read cbfs file header\n");
++			return EXIT_FAILURE;
++	  }
++
++		file.len = ntohl(file.len);
++		file.type = ntohl(file.type);
++		file.attributes_offset = ntohl(file.attributes_offset);
++		file.offset = ntohl(file.offset);
++
++		if (verbose) {
++			printf("File magic             : %.8s\n", file.magic);
++			printf("File len               : %x\n", file.len);
++			printf("File type              : %x\n", file.type);
++			printf("File attributes_offset : %x\n", file.attributes_offset);
++			printf("File offset            : %x\n", file.offset);
++		}
++
++		if (strncmp((char *)file.magic, CBFS_FILE_MAGIC, 8) != 0) {
++			break;
++		}
++
++		size_t name_size = file.offset - sizeof(struct cbfs_file);
++		char *name = calloc(name_size, sizeof(char));
++		if (read(fd, name, name_size) < 0) {
++			fprintf(stderr, "Failed to read cbfs file name\n");
++			return EXIT_FAILURE;
++		}
++		if (verbose) {
++			printf("File name              : '%s'\n", name);
++		}
++
++		if (do_list && file.type == 0x50) {
++			printf("%s\n", name);
++		}
++		if (do_read && file.type == 0x50 &&
++			strncmp(name, filename, name_size) == 0)
++		{
++			uint64_t file_off = off + file.offset;
++			if (verbose) {
++				printf("Seeking to %lx\n-------- Start Data\n", file_off);
++			}
++			if (lseek(fd, file_off, SEEK_SET) == -1) {
++				fprintf(stderr, "Failed to seek to file data at %lx\n", file_off);
++				return EXIT_FAILURE;
++			}
++
++			char *file_data = malloc(file.len);
++			if (read(fd, file_data, file.len) < 0) {
```

Comment:
Verify file_data was allocated correctly.. A corrupted image could cause you to try and allocate GBs of data which would fail.